extensions/amp-video/0.1/amp-video.js

```diff
@@ -86,8 +85,112 @@ class AmpVideo extends AMP.BaseElement {
     /** @private {!../../../src/mediasession-helper.MetadataDef} */
     this.metadata_ = EMPTY_METADATA;
 
-    /** @private @const {!Array<!UnlistenDef>} */
-    this.unlisteners_ = [];
+    /** @private {?./mixins.AmpVideoMixinDef} */
+    this.mixin_ = null;
+
+    /** @private {?./mixins.PlaybackMixin} */
+    this.playbackMixin_ = null;
+
+    /** @private {boolean} */
+    this.isConnected_ = false;
+  }
+
+  /** @return {boolean} @visibleForTesting */
+  // TODO(alanorozco): Use mediapool service to determine mixin.
+  static isInsideStory(element) {
+    // Small optimization to short-circuit on amp-story-grid-layer instead of
+    // amp-story. Validation defines that all content inside a Story should be
+    // inside an <amp-story-grid-layer>.
+    return !!closestByTag(element, 'amp-story-grid-layer');
+  }
+
+  /** @override */
+  firstAttachedCallback() {
+    this.isConnected_ = true;
+  }
+
+  /**
+   * @param {boolean=} isLaidOut
+   * @return {!./mixins.AmpVideoMixinDef}
+   */
+  getMixin_(isLaidOut = false) {
+    // Use a ascendant-agnostic mixin before we know whether the component
+    // descends from a mediapool-bound element. Determine whether it comes from
+    // a mediapool on `layoutCallback`.
+    //
+    // Note that this can be determined by SSR, via the `i-amphtml-mediapool`
+    // and `i-amphtml-not-mediapool` classnames. See `determineMixinFromSsr_`.
+
+    if (this.mixin_ &&
+        (this.mixin_.isIntermediate() && !isLaidOut ||
+        this.mixin_.controlsPlayback())) {
+      return this.assertMixin_();
+    }
+
+    const ctor = this.determineMixin_(isLaidOut);
+    const mixin = new ctor(this);
+
+    if (mixin.controlsPlayback()) {
+      this.playbackMixin_ = mixin;
+    }
+
+    if (this.mixin_ !== null) {
+      dev().assert(this.mixin_.isIntermediate());
+
+      const intermediateMixin = /** @type {!IntermediatePropagationMixin} */ (
+        dev().assert(this.mixin_));
+
+      intermediateMixin.transferState(mixin);
+    }
+
+    this.mixin_ = mixin;
+
+    return this.assertMixin_();
+  }
+
+  /** @private @return {!./mixins.AmpVideoMixinDef} */
+  assertMixin_() {
+    return /** @type {!./mixins.AmpVideoMixinDef} */ (
+      dev().assert(this.mixin_));
+  }
+
+  /**
+   * @param {boolean} isLaidOut
+   * @return {./mixins.AmpVideoMixinConstructorDef}
+   * @private
+   */
+  determineMixin_(isLaidOut) {
+    const ctorFromCacheDetermination = this.determineMixinFromSsr_();
+
+    if (ctorFromCacheDetermination) {
+      return dev().assert(ctorFromCacheDetermination);
+    }
+
+    const viewer = Services.viewerForDoc(this.getAmpDoc());
+
+    if (viewer.getVisibilityState() != VisibilityState.PRERENDER &&
```

Comment:
add comments on why `viewer.getVisibilityState() != VisibilityState.PRERENDER` is needed. Not clear to me